aten/src/ATen/TensorIndexing.h

```diff
@@ -463,35 +464,38 @@ static inline Tensor handleDimInMultiDimIndexing(
     Tensor result = prev_dim_result;
     const Tensor& tensor = index.tensor();
     auto scalar_type = tensor.scalar_type();
-    if (tensor.dim() == 0 &&
-        at::isIntegralType(scalar_type, /*includeBool=*/true)) {
-      if (scalar_type != at::kByte && scalar_type != at::kBool) {
-        result = impl::applySelect(
-            result,
-            *dim_ptr,
-            tensor.item<int64_t>(),
-            real_dim,
-            original_tensor_device,
-            prev_dim_result_sizes);
-      } else {
-        result = result.unsqueeze(*dim_ptr);
-        if (scalar_type == at::kBool) {
-          impl::recordTensorIndex(
-              impl::boolToIndexingTensor(
-                  result, tensor.item<bool>() != 0, original_tensor_device),
-              outIndices,
-              dim_ptr);
-        } else {
-          impl::recordTensorIndex(
-              impl::boolToIndexingTensor(
-                  result, tensor.item<uint8_t>() != 0, original_tensor_device),
-              outIndices,
-              dim_ptr);
-        }
-      }
-    } else {
+    if (tensor.dim() != 0) {
       impl::recordTensorIndex(tensor, outIndices, dim_ptr);
+      return result;
+    }
+    if (scalar_type == at::kByte) {
+      result = result.unsqueeze(*dim_ptr);
+      impl::recordTensorIndex(
+          impl::boolToIndexingTensor(
+              result, tensor.item<uint8_t>() != 0, original_tensor_device),
+          outIndices,
+          dim_ptr);
+      return result;
+    } else if (scalar_type == at::kBool) {
+      result = result.unsqueeze(*dim_ptr);
+      impl::recordTensorIndex(
+          impl::boolToIndexingTensor(
+              result, tensor.item<bool>() != 0, original_tensor_device),
+          outIndices,
+          dim_ptr);
+      return result;
+    } else if (
+        at::isIntegralType(scalar_type, /*includeBool=*/false) &&
+        tensor.device() == at::kCPU && !at::isTensorSubclassLike(tensor)) {
+      return impl::applySelect(
+          result,
+          *dim_ptr,
+          tensor.item<int64_t>(),
+          real_dim,
+          original_tensor_device,
+          prev_dim_result_sizes);
```

Comment:
I think we want to look for an alternative method. This PR changes the semantics of the following code:
```py
x = torch.arange(10)
y = torch.tensor(2)
def f(x, y):
    return x[y]

gm = make_fx(f)(x, y)
z = gm(x, y)
```
Before this PR, z is a view of x. After this PR, we end up invoking advanced indexing (torch.ops.aten.index.Tensor) instead of regular indexing, which ends up creating a new Tensor instead of returning us a view

Reply:
IMO, that's reasonable. To be clear, it *only* does that if the tensor is currently a cuda tensor. I'd argue that the current semantics are not great, actually, since it actually incurs a cuda sync.

Reply:
Also, we theoretically aren't able to avoid the CUDA D2H sync for x[y]. At some point we need to know what `y` is so that we can construct the view.

Reply:
You can't just wish away the BC semantics lol.

@Chillee, would it help if dynamic shapes is able to trace through the sync and then we "repair" it post facto to not sync?

Reply:
>:(

Reply:
@ezyang My thinking is that we should just always make it not sync. Cuda synchronizations are almost always bad/something to be avoided, and in cases where there's no particularly good reason for it to sync, we should err on the side of not syncing.

Dynamic shapes able to trace through the sync would be .... ok, but I think if we need to do that I'll probably just require folks to explicitly use an indexing operator (which is kinda dumb).

Reply:
Actually, I can work around this pretty easily by just making the scalars 1-dim instead of 0-dim :/ But I still think we should change the semantics here.

Reply:
I put up an issue for discussing this change: https://github.com/pytorch/pytorch/issues/105641

Reply:
You might want to look into using `index_select` or `index_copy` for these things instead, rather than changing the semantics (which I agree are not great, but well). See the `x.ndim == 0` case of
https://github.com/pytorch/pytorch/blob/1ea153a11d3011b90cdac1f9977889988a0c981f/torch/_refs/__init__.py#L3621-L3637

Reply:
Should there be an explicit-semantics method of calling `x[idx]` more explicitly, e.g. reintroduce:

- https://github.com/pytorch/pytorch/issues/41625
- and maybe create more explicit bindings instead of only `tensor.__getitem__` allowing to express various special semantics when needed